src/shell.cpp

```diff
@@ -0,0 +1,243 @@
+    }
+}
+
+/**
+ * Unsets an environment variable.
+ * @param args The command arguments.
+ */
+void UnsetEnvironmentVariable(const std::vector<std::string>& args) {
+    if (args.size() == 1) {
+        environmentVariables.erase(args[0]);
+    } else {
+        std::cerr << "Invalid arguments for 'unset' command." << std::endl;
+    }
+}
+
+/**
+ * Expands environment variables in the command arguments.
+ * @param args The command arguments.
+ */
+void ExpandEnvironmentVariables(std::vector<std::string>& args) {
+    for (std::string& arg : args) {
+        if (arg.empty())
+            continue;
+
+        if (arg[0] != '$' && arg[0] != '{')
+            continue;
+
+        std::string varName = arg.substr(1);
+        if (varName.empty())
+            continue;
+
+        if (varName[0] == '{' && varName.back() == '}') {
+            varName = varName.substr(1, varName.size() - 2);
+        }
+
+        auto it = environmentVariables.find(varName);
+        if (it != environmentVariables.end()) {
+            arg = it->second;
+        } else {
+            std::cerr << "Environment variable not found: " << varName << std::endl;
+        }
+    }
+}
+
+/**
+ * Handles the execution of built-in commands.
+ * @param command The command to handle.
+ * @param args The command arguments.
+ * @param backgroundProcesses The list of background process IDs.
+ * @return True if the command is a built-in command and handled successfully, false otherwise.
+ */
+bool HandleBuiltInCommands(const std::string& command, const std::vector<std::string>& args, std::vector<pid_t>& backgroundProcesses) {
+    if (command == "myjobs") {
+        PrintBackgroundJobs(backgroundProcesses);
+        return true;
+    } else if (command == "set") {
+        SetEnvironmentVariable(args);
+        return true;
+    } else if (command == "unset") {
+        UnsetEnvironmentVariable(args);
+        return true;
+    }
+    return false;
+}
+
+/**
+ * Processes a command line.
+ * @param commandLine The command line to process.
+ * @param backgroundProcesses The list of background process IDs.
+ */
+void ProcessCommandLine(const std::string& commandLine, std::vector<pid_t>& backgroundProcesses) {
+    std::vector<std::string> tokens = Split(commandLine, ' ');
+    if (tokens.empty()) {
+        return;
+    }
+
+    std::string command = tokens[0];
+    std::vector<std::string> args(tokens.begin() + 1, tokens.end());
+
+    bool runInBackground = IsBackgroundCommand(args);
+    if (runInBackground) {
+        RemoveBackgroundSymbol(args);
+    }
+
+    ExpandEnvironmentVariables(args);
+
+    if (HandleBuiltInCommands(command, args, backgroundProcesses)) {
+        return;
+    }
+
+    ExecuteCommand(command, args, runInBackground);
+
+    if (runInBackground) {
+        pid_t pid = fork();
+        if (pid < 0) {
+            std::cerr << "Failed to create child process." << std::endl;
+        } else if (pid > 0) {
+            backgroundProcesses.push_back(pid);
```

Comment:
Who is going to wait on the background jobs so it checks their status? Do you know what happens to processes whose parents don't "wait" for them?